contracts/core/Core.sol

```diff
@@ -416,6 +415,55 @@ contract Core is
         return newSetTokenAddress;
     }
 
+    /**
+     * Function to convert Set Tokens into underlying components
+     *
+     * @param _setAddress    The address of the Set token
+     * @param _quantity        The number of tokens to redeem
+     */
+    function redeem(
+        address _setAddress,
+        uint _quantity
+    )
+        public
+        isValidSet(_setAddress)
+        isNonZero(_quantity)
+        isNaturalUnitMultiple(_quantity, _setAddress)
+    {
+        uint naturalUnit = ISetToken(_setAddress).naturalUnit();
+
+        // Transfer the underlying tokens to the corresponding token balances
+        address[] memory components = ISetToken(_setAddress).getComponents();
+        uint[] memory units = ISetToken(_setAddress).getUnits();
+        for (uint16 i = 0; i < components.length; i++) {
+            address currentComponent = components[i];
+            uint currentUnit = units[i];
+
+            uint tokenValue = calculateTransferValue(
+                currentUnit,
+                naturalUnit,
+                _quantity
+            );
+
+            // Decrement the Set amount
+            IVault(vaultAddress).decrementTokenOwner(
+                _setAddress,
+                currentComponent,
+                tokenValue
+            );
+
+            // Increment the component amount
+            IVault(vaultAddress).incrementTokenOwner(
+                msg.sender,
+                currentComponent,
+                tokenValue
+            );
+        }
+
+        // Burn the Set token (thereby decrementing the SetToken balance)
+        ISetToken(_setAddress).burn(msg.sender, _quantity);
+    }
```

Comment:
I wonder if its useful to emit a `Redeem` event.

Reply:
+1. I didn't emit any events for issue, though now that I think of it, it might be useful to emit an event for when a Transfer was required for a token.

Reply:
Is it enough for the `emit Transfer(_from, address(0), _quantity);` event in `burn`?